### ISSUE TITLE
Missing examples validating data used to make query or mutation

### ISSUE THREAD

**Issue by kristianmandrup** [aws-amplify/amplify-js#1561]
Been searching for any documentation on how to best add validation to params/data used in GraphQL queries and mutations. Would it be best done from the client app such as React as part of an action or reducer (f.ex in Redux os similar store). What is the recommended approach? I can see there is now some limited validation support in the Velocity templates as well.
Please provide some guidance/examples in the docs.
Thanks :)

**Comment by kristianmandrup** [comment]
Best example I've found so far is [here](https://gyandeeps.com/aws-appsync-graphql/)

I guess it could be tweaked to sth like:

```js
class App extends React.Component {
    render() {
        return (
            <Query query={GET_ORDERS}>
                {({ data }) => (
                    <Mutation mutation={ADD_TODO}>
                        {(addTodo) => (
                            <>
                                <div>{JSON.stringify(data)}</div>
                                <button
                                    onClick={() =>
                                        const {data, error} = validate.todo({
                                            variables: { text: "gyandeep" }
                                        })
                                        (data && addTodo(data)) ||  (error && dispatch(error))
                                    }
                                >
                                    add
                                </button>
                            </>
                        )}
                    </Mutation>
                )}
            </Query>
        );
    }
}
```

Would this be a good approach?

**Comment by kristianmandrup** [comment]
Found another good example [here](https://docs.aws.amazon.com/appsync/latest/devguide/building-a-client-app-react.html)

I guess the method such as `onEdit: (post) => { props.mutate({... ` could include a validation step before calling mutate on props?

```js
const AllPostsWithData = compose(
    graphql(AllPostsQuery, {
        options: {
            fetchPolicy: 'cache-and-network'
        },
        props: (props) => ({
            posts: props.data.allPost && props.data.allPost.posts,
        })
    }),
    graphql(DeletePostMutation, {
        props: (props) => ({
            onDelete: (post) => props.mutate({
                variables: { id: post.id, expectedVersion: post.version },
                optimisticResponse: () => ({ deletePost: { ...post, __typename: 'Post' } }),
            })
        }),
        options: {
            refetchQueries: [{ query: AllPostsQuery }],
            update: (proxy, { data: { deletePost: { id } } }) => {
                const query = AllPostsQuery;
                const data = proxy.readQuery({ query });

                data.allPost.posts = data.allPost.posts.filter(post => post.id !== id);

                proxy.writeQuery({ query, data });
            }
        }
    }),
    graphql(UpdatePostMutation, {
        props: (props) => ({
            onEdit: (post) => {
                props.mutate({
                variables: { ...post, expectedVersion: post.version },
                optimisticResponse: () => ({ updatePost: { ...post, __typename: 'Post', version: post.version + 1 } }),
                })
            }
        }),
        options: {
            refetchQueries: [{ query: AllPostsQuery }],
            update: (dataProxy, { data: { updatePost } }) => {
                const query = AllPostsQuery;
                const data = dataProxy.readQuery({ query });

                data.allPost.posts = data.allPost.posts.map(post => post.id !== updatePost.id ? post : { ...updatePost });

                dataProxy.writeQuery({ query, data });
            }
        }
    })
    )(AllPosts);

    const NewPostWithData = graphql(NewPostMutation, {
    props: (props) => ({
        onAdd: post => props.mutate({
            variables: post,
            optimisticResponse: () => ({ addPost: { ...post, __typename: 'Post', version: 1 } }),
        })
    }),
    options: {
        refetchQueries: [{ query: AllPostsQuery }],
        update: (dataProxy, { data: { addPost } }) => {
            const query = AllPostsQuery;
            const data = dataProxy.readQuery({ query });

            data.allPost.posts.push(addPost);

            dataProxy.writeQuery({ query, data });
        }
    }
})(AddPost);
```

From the project code referenced/used.

```js
export default class AddPost extends Component {
  // ....

    handleAdd = () => {
        const { id, title, author } = this.state;

        this.setState(this.getInitialState(), () => {
            this.props.onAdd({ id, title, author });
        });
    }

    render() {
        return (
            <fieldset >
                <legend>Add new Post</legend>
                <div>
                    <label>ID<input type="text" placeholder="ID" value={this.state.id} onChange={this.handleChange.bind(this, 'id')} /></label>
                </div>
                <div>
                    <label>Title<input type="text" placeholder="Title" value={this.state.title} onChange={this.handleChange.bind(this, 'title')} /></label>
                </div>
                <div>
                    <label>Author<input type="text" placeholder="Author" value={this.state.author} onChange={this.handleChange.bind(this, 'author')} /></label>
                </div>
                <div>
                    <button onClick={this.handleAdd}>Add new post</button>
                    <button onClick={this.handleCancel}>Cancel</button>
                </div>
            </fieldset>
        );
    }
}
```

**Comment by jordanranz** [comment]
Hey @kristianmandrup. Did you figure out a good approach for your app?

**Comment by kristianmandrup** [comment]
I created several libraries published on `npmjs` which can take a GraphQL schema and generate a validator for it.

- [json-schema-to-yup](https://www.npmjs.com/package/json-schema-to-yup) also for GraphQL schema
- [graphql-constraint-directive-plus](https://www.npmjs.com/package/graphql-constraint-directive-plus)

Haven't gone back to using `amplify-js` though. Found it a little too "immature" at the time. 
Might give it a new go sometime this year!
